### PR TITLE
[@types/puppeteer] Add definitions for Puppeteer v4 and v5.4

### PR DESCRIPTION
**Pull request by cameronhunter**
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/puppeteer/puppeteer/releases
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## pa11y ([<kbd>unpkg</kbd>](https://unpkg.com/browse/pa11y@latest/))
was missing the following properties:
1. defaults
1. allowedStandards
## puppeteer ([<kbd>unpkg</kbd>](https://unpkg.com/browse/puppeteer@latest/))
was missing the following properties:
1. product
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 1f49a772605937cde182483f107de5c67cd2a043
</p>

**Review by azerella — APPROVED** [review]
LGTM ♥️

**Comment by adamzerella** [comment]
@cameronhunter Just checking you're aware that the pipeline benchmark is failing

**Comment by cameronhunter** [comment]
@adamzerella thanks, I wasn't sure what to do about it failing. I'll read the troubleshooting guide tomorrow.

**Comment by alvis** [comment]
@cameronhunter First thanks a lot for the update!!! It has been long overdue! Meanwhile, do you think you have time to add the missing `waitForTimeout` method as well? Something #48989 has been trying.

File: types/puppeteer/index.d.ts
Comment (cameronhunter) [inline types/puppeteer/index.d.ts:1331]: @alvis – `waitForTimeout` is already in this PR.

**Comment by cameronhunter** [comment]
@adamzerella Is there anything else I need to do with this branch or is it just waiting for approval from another maintainer?

**Review by azerella — APPROVED** [review]
❤️ it

**Comment by cameronhunter** [comment]
Ready to merge

**Comment by christian-bromann** [comment]
@cameronhunter this has broken our build due to adding

```ts
export interface CustomQueryHandler {
    queryOne?: (element: Element | Document, selector: string) => Element | null;
    queryAll?: (element: Element | Document, selector: string) => Element[] | NodeListOf<Element>;
}
```

Where `Document` has no definition, causing this TS error:

```
../../../packages/webdriverio/node_modules/@types/puppeteer/index.d.ts:2361:36 - error TS2304: Cannot find name 'Documnt'.
                                                                                                                    
2361     queryOne?: (element: Element | Document, selector: string) => Element | null
                                        ~~~~~~~~
../../../packages/webdriverio/node_modules/@types/puppeteer/index.d.ts:2362:36 - error TS2304: Cannot find name 'Documnt'.
                                                                                                                    
2362     queryAll?: (element: Element | Document, selector: string) => Element[] | NodeListOf<Element>;
                                        ~~~~~~~~
```

Any advice or shall we remove `Document`?

**Comment by cameronhunter** [comment]
@christian-bromann `Document` is part of `dom` lib so I'm surprised that you're seeing failures. Can you confirm that you have a `tsconfig.json` file that contains:

```json
{
  "compilerOptions": {
    "lib": ["dom"]
  }
}
```

**Comment by christian-bromann** [comment]
@cameronhunter thanks! I tried that and failed initially because I might have applied it in a wrong tsconfig, someone updated [their PR](https://github.com/webdriverio/webdriverio/pull/6032/files#r518199984) to include this fix. Thank you for the response.

**Comment by tyrak** [comment]
@cameronhunter With this PR, I am getting an error by tsc related to `TimeoutError`. Specifically, in many places I have code like this:

    try {
        // do some puppeteer stuff here
    }
    catch (err) {
        if (err instanceof puppeteer.errors.TimeoutError)
            // do something
        else
            // do something else
    }

and now I get the following error:

    error TS2359: The right-hand side of an 'instanceof' expression must be of type 'any' or of a type assignable to the 'Function' interface type.

**Comment by tyrak** [comment]
If you change the definition of `errors` to

    export namespace errors {
        class TimeoutError extends CustomError {}
    }

this fixes the issue for us. I see you are also exporting the object `errors.puppeteerErrors`. Are you sure that this object is actually exported by puppeteer? In node repl I get this:

    > require("puppeteer").errors.puppeteerErrors
    undefined

**Pull request merged**
